### ISSUE TITLE
map + pow to successor + mul

### ISSUE THREAD

**Issue by wuyudi** [rust-lang/rust-clippy#9751]
### What it does

Use successor instead of map of pow to calculate pow, etc.

### Lint Name

map_pow_to_successor

### Category

perf

### Advantage

- better perf, origin:  $O(n * \log(n))$   , if use exponentiating by squaring. Now:  $O(n)$

### Drawbacks

Longer code.

### Example

```rust
fn foo(y: u32) -> Vec<u32> {
    (0..).map(|i| y.pow(i)).take(5).collect()
}
fn main() {
    println!("{:?}", foo(7));
}
```

Could be written as:

```rust
use std::iter::successors;
fn foo(y: u32) -> Vec<u32> {
    successors(Some(1), |x| Some(y * x)).take(5).collect()
}
fn main() {
    println!("{:?}", foo(7));
}
```

**Comment by Centri3** [comment]
`successors` is actually ~2x slower here (another case of big O notation failing us!), and the assembly output is not at all reassuring. While not a good way to measure performance it's ~2x larger as well mostly due to checks that would be optimized out in `map`, which means a good chunk less cache hits and an absolute TON of branches. While I do think the code is nicer this is not at all `perf`.

**Comment by Centri3** [comment]
Opened: https://github.com/rust-lang/rust/issues/113055
